### ISSUE TITLE
support for adblock?

### ISSUE THREAD

**Issue by MakuSensei** [jiahaog/nativefier#177]
### Description

Hello, you could block some websites advertising
### Specifications
- Version of Nativefier (run `$ nativefier --version`): `v6.13.0`
- Version of Node.js (run `$ node --version`): `v4.4.0`
- OS: `windows 10`
- Architecture `x64`
- Stack trace from the error message (if any)

```
<Stack trace here>
```

**Comment by jiahaog** [comment]
This might be possible with an injected script, as mentioned in #3. Would you like to work on a pull request?

**Comment by eladnava** [comment]
@jiahaog I'd love to give it a try if that's okay with you. =)

**Comment by jiahaog** [comment]
Sorry for the late reply! Sure, that'll be great!

**Comment by eladnava** [comment]
@jiahaog Do you have any pointers on how this should work? It doesn't look like Electron supports adding non-DevTools extensions (such as Adblock).

**Comment by jiahaog** [comment]
Would it be possible to inject some sort of javascript adblocker prior to loading the page? How about trying to first using [`--inject`](https://github.com/jiahaog/nativefier/blob/development/docs/api.md#inject) to test the adblock functionality prior to integrating it to the app itself?

**Comment by eladnava** [comment]
I looked all over the interwebs for a JavaScript AdBlocker that can be embedded - I couldn't find such a beast. They only exist as extensions, because what site is going to include an AdBlocker for its own ads?

It appears the only available course of action is to port one of the extensions for use with Electron because they all heavily rely on the chrome.* extension APIs. It appears Electron implements some of the functionality offered to Chrome extensions, but via a slightly different API interface.

AdBlock Plus for Chrome is open source and seems like a good candidate for this:
https://github.com/adblockplus/adblockpluschrome

I'm not sure I'll have the time to work on it, though -- it's a pretty big project, actually. Anyone else willing to give it a shot?

**Comment by jiahaog** [comment]
@eladnava You're right, it's probably going to involve porting the chrome extension to electron, and will take alot of work.

Apparently it is possible to [add chrome dev extensions to electron](https://www.xplatform.rocks/2016/07/02/using-augury-inside-of-your-electron-apps/). I'm not too sure, but would that be a solution for this?

**Comment by eladnava** [comment]
@jiahaog Unfortunately, that won't be enough, because Adblock relies on the `chrome.*` extension APIs, which are simply not implemented in Electron (yet):
https://github.com/electron/electron/issues/1498#issuecomment-96995155

**Comment by matthewdias** [comment]
could do something like this maybe http://www.javascriptsource.com/cookies/remove-ads.html

**Comment by eladnava** [comment]
@matthewdias unfortunately that script needs to be configured with each advertisement's URL in order to work. It's for websites that want to reward the user for clicking a specific ad by hiding it in the future.

**Comment by eladnava** [comment]
A manageable workaround is to just block ad network hosts using the OS `hosts` file. Since a nativefied website will typically serve the same source of ads (from the same ad network), it is pretty easily blockable.

**Comment by ronjouch** [comment]
Closing this as electron [*"don't have any plan to support Chrome extension APIs"*](https://github.com/electron/electron/issues/1498#issuecomment-96995155). Workarounds don't seem very satisfactory :-/ , but as @eladnava mentioned, system-wide blocking via hosts file (e.g. [Dan Pollock's hosts file](http://someonewhocares.org/hosts/)) can be a workaround. ¯\_(ツ)_/¯

**Comment by remusao** [comment]
Hi @jiahaog, @ronjouch,

I just stumbled upon this issue and thought I would mention this JavaScript [adblocker library](https://github.com/cliqz-oss/adblocker) which I am maintaining. It has first-class support for electron and can be integrated with only a [few lines of code](https://github.com/cliqz-oss/adblocker/tree/master/packages/adblocker-electron#usage).

The only challenge I see is that to get full blocking capabilities (like adblocker extensions), it needs to register a preload script and I'm not sure how that would work with `electron-packager`. I guess it might require to copy the preload script at build-time so that it's available (otherwise it will still perform network filtering, which is usually enough to block most ads).

Best,
Rémi

**Comment by ronjouch** [comment]
> I just stumbled upon this issue and thought I would mention this JavaScript [adblocker library](https://github.com/cliqz-oss/adblocker) which I am maintaining. It has first-class support for electron and can be integrated with only a [few lines of code](https://github.com/cliqz-oss/adblocker/tree/master/packages/adblocker-electron#usage).
>
> The only challenge I see is that to get full blocking capabilities (like adblocker extensions), it needs to register a preload script and I'm not sure how that would work with `electron-packager`. I guess it might require to copy the preload script at build-time so that it's available (otherwise it will still perform network filtering, which is usually enough to block most ads).

@remusao cool! PR welcome, and I don't have the bandwidth to help, but I'm reopening this issue (I closed it because Electron doesn't support Chrome extension APIs, but what you offer would be a nice alternative) to help give it visibility, maybe others will be able to help 🙂.

**Comment by remusao** [comment]
Thanks for the super fast answer @ronjouch. I actually can make a simple POC of this if that helps. I would still need help though, to package the custom preload script with the app. Not sure how that could work?

**Comment by My6UoT9** [comment]
@remusao should be easy now to preload js, https://github.com/jiahaog/nativefier/blob/master/docs/api.md#inject

**Comment by remusao** [comment]
@My6UoT9 Thanks for the suggestion, let's maybe have a look again once the refactor to TypeScript by @ronjouch is merged?

**Comment by TheCleric** [comment]
After I get done with my PR for issue #29 I will be moving on to work on this.

I am thinking about following along the path laid out in this thread:

https://discuss.atom.io/t/resolved-adblock-for-electron-app/39396/6

And possibly using this node library mentioned therein:

https://www.npmjs.com/package/abp-filter-parser

Let me know if anyone has any questions/comments or would be willing to help.

**Comment by remusao** [comment]
Hey @TheCleric,

As mentioned in my previous [comment here](https://github.com/jiahaog/nativefier/issues/177#issuecomment-526934864), I would be willing to help. I want to point out that the [adb-filter-parser](https://www.npmjs.com/package/abp-filter-parser) library is not maintained anymore (no update is 4 years).

Best,

**Comment by TheCleric** [comment]
@remusao thanks for pointing that out about that library. I missed that. When I'm ready to do some work on this I'll research some newer options, and ping you again. Thanks!

**Comment by garywu520** [comment]
Please refer to this project, I have completed the basic test, welcome to improve it together. https://github.com/garywu520/webapps-ads-removal/blob/main/README.md

**Comment by auronpl315** [comment]
I'm new, how do I add Terminal to Adblock (nativefier Electron)?

<img width="760" alt="Screenshot 2023-04-19 at 18 01 53" src="https://user-images.githubusercontent.com/72985338/233134367-32632a58-b615-48bc-b838-2ba4231a08a3.png">


<img width="1280" alt="Screenshot 2023-04-19 at 18 00 09" src="https://user-images.githubusercontent.com/72985338/233133975-6a75a345-71ab-4e57-af9a-e275cac3083d.png">

**Comment by cwlovecs** [comment]
We can use `--inject ad.css`

ad.css
```css
.some-ad {
	display: none
}
```
